### PR TITLE
Implementation of the municipality map using highcharts

### PR DESCRIPTION
**Pull request by VWSCoronaDashboard1**
## Summary

Rudimentary implementation of the municipality map using highcharts

## Motivation

Feature request

## Detailed design

The map is constructed using two series, one for the country outline and the other for the municipal outlines.
Both are loaded from the static-json directory in the public folder (not sure if this is the correct way of going about it, let me know).
The actual metric data is loaded from json/municipality-data.json, this is a data file that should contain an array of municipal objects in this shape:

```
{
  "Date_of_report": "2020-03-13 10:00:00",
  "Municipality_code": "GM0003",
  "Municipality_name": "Appingedam",
  "Province": "Groningen",
  "Total_reported": 0,
  "Hospital_admission": 0,
  "Deceased": 0
}
``` 

the useMunicipalityData hook loads this file and extracts only those objects with the newest Date_of_report (which should be the current day, but if not then it'll just take the latest, that really depends on the delivered data). The hook then also creates a 'value' property on each item whose value is filled by the given metricName property (Total_reported, Hospital_admission or Deceased).

These three datasources are then fed into the highchart config where the geojson data is cross-referenced with the metric data using the gemcode -> Municipality_code link.

The Chart component itself has three props:
selected?: { id: string }; // this pre-selects the indicated municipality on the map
setSelection: (item: { id: string }) => void; // this method is called after the user click on the map
metric: TMunicipalityMetricName; // this indicates which metric is supposed to be rendered as the chloropleth (Total_reported, Hospital_admission or Deceased)

## Drawbacks

n/a

## Alternatives

n/a

## Unresolved questions

For now I;ve only just added the component to the region page. Its not hooked up to any of the KPIs or whatever. Its also not using a correct gradient for the cholorpleth yet. But functionaly this is pretty much what was expected I believe.

### CODE REVIEW

File: src/components/mapChart/index.tsx [hunk 1]
```diff
@@ -1,135 +1,161 @@
...
+        maxColor: '#ff0000',
+      },
+      legend: {
+        enabled: false,
+      },
+      tooltip: {
+        backgroundColor: '#FFF',
+        borderColor: '#01689B',
+        borderRadius: 0,
+        formatter: function (this: { point: TMunicipalityPoint }) {
```
Comment (VWSCoronaDashboard4) [inline src/components/mapChart/index.tsx:82]: Add the return type to the function

File: src/components/mapChart/index.tsx [hunk 2]
```diff
@@ -1,135 +1,161 @@
...
+      metric,
+    ]
+  );
+
+  const ref = useRef<any>();
+
+  useEffect(() => {
+    const chart = ref?.current?.chart;
+    if (selected && chart) {
+      if (!chart) {
```
Comment (VWSCoronaDashboard4) [inline src/components/mapChart/index.tsx:146]: chart is already truthy due to the check by the previous if statement

File: src/components/mapChart/index.tsx [hunk 3]
```diff
@@ -1,135 +1,161 @@
...
+        return;
+      }
+      const point = chart?.series[0].points.find(
+        (p: any) => p.gemcode === selected.id
```
Comment (VWSCoronaDashboard4) [inline src/components/mapChart/index.tsx:150]: any could be replaced by actual type

File: src/pages/regio/index.tsx [hunk 4]
```diff
@@ -165,15 +163,20 @@ const Regio: FCWithLayout<RegioProps> = (props) => {
 
             <div className={styles['map-container']}>
               <div className={styles['safety-region-header']}>
-                <p>{text.your_safety_region}</p>
```
Comment (VWSCoronaDashboard4) [inline src/pages/regio/index.tsx:168]: Develop was not properly merged here, the translations should be used

File: src/utils/useMunicipalityData.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      return [];
+    }
+
+    const max = Math.min(
+      ...data.map((item) => new Date(item.Date_of_report).getTime())
+    );
+
+    const filteredData = data
+      .filter(
+        (item) =>
```
Comment (VWSCoronaDashboard4) [inline src/utils/useMunicipalityData.ts:31]: add type to the item

File: src/utils/useMunicipalityData.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+): MunicipalityData[] {
+  const { data } = useSWR<MunicipalityData[]>('/json/municipality-data.json');
+
+  return useMemo(() => {
+    if (!data) {
+      return [];
+    }
+
+    const max = Math.min(
+      ...data.map((item) => new Date(item.Date_of_report).getTime())
```
Comment (VWSCoronaDashboard4) [inline src/utils/useMunicipalityData.ts:26]: add type to the item

File: src/components/mapChart/index.tsx (hunk 2 above)
Comment (VWSCoronaDashboard1) [reply src/components/mapChart/index.tsx:146]: I'm not sure I follow. IMHO, chart could be undefined when either ref or current is undefined, or when current IS defined but the chart isn't fully initialized yet.

File: src/pages/regio/index.tsx (hunk 4 above)
Comment (VWSCoronaDashboard1) [reply src/pages/regio/index.tsx:168]: I've created this branch based on 'next', so I haven't merged anything from develop.

File: src/components/mapChart/index.tsx (hunk 3 above)
Comment (VWSCoronaDashboard1) [reply src/components/mapChart/index.tsx:150]: Actually, typing this strictly will get messy. The Highcharts type definitions don't provide a generic Series type that allows for the point property to be typed. So, currently its just typed to the 'Series.Point' type. I've typed the chart variable to Highcharts.Chart which expects the chart.series.point property to be just that. So typing p in the map will trigger a compile error.
I don't really see a neat way of fixing this properly, so if you have a suggestion, I'm all ears.

File: src/components/mapChart/index.tsx [hunk 7]
```diff
@@ -1,135 +1,161 @@
...
+    FeatureCollection<MultiPolygon, MunicipalityProperties>
+  >('/static-json/municipalities-outline.geojson');
+
+  const municipalityData = useMunicipalityData(metric);
+  const [min, max] = useExtent(
+    municipalityData,
+    (item: MunicipalityData) => item[metric]
+  );
+
+  const mapOptions = useMemo(
```
Comment (VWSCoronaDashboard3) [inline src/components/mapChart/index.tsx:49]: In order to get the correct typings for the options you need `useMemo<Highcharts.Options>`

File: src/utils/useExtent.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * This hook returns the minimum and maximum values for the given array.
+ * Optionally a predicate can be passed in to extract the number values from the given array.
+ *
+ * @param collection the given array
+ * @param predicate the optional predicate
+ */
+export default function useExtent(
+  collection: any[],
+  predicate?: (item: any) => number
```
Comment (VWSCoronaDashboard3) [inline src/utils/useExtent.ts:12]: This might be a bit more correct.
```suggestion
  predicate?: (item: Record<string, unknown>) => number
```

Comment (VWSCoronaDashboard3) [reply src/utils/useExtent.ts:12]: Or if arrays are supported as well.

```ts
Record<string, unknown> | Array<unknown>
```

**Review by VWSCoronaDashboard3 — COMMENTED** [review]
Looks good! I like that functionality has been spit up into hooks.

File: src/components/mapChart/index.tsx [hunk 9]
```diff
@@ -79,14 +79,22 @@ const MunicipalityMap: React.FC<IProps> = ({
...
-        formatter: function (this: { point: TMunicipalityPoint }) {
-          const { Province, Municipality_name } = this.point;
+        formatter: function (
+          this: TooltipFormatterContextObject
+        ):
+          | false
+          | string
+          | Array<string | null | undefined>
+          | null
+          | undefined {
```
Comment (VWSCoronaDashboard3) [inline src/components/mapChart/index.tsx:89]: Interesting 😅

File: src/components/mapChart/index.tsx (hunk 7 above)
Comment (VWSCoronaDashboard1) [reply src/components/mapChart/index.tsx]: Already done

**Review by VWSCoronaDashboard3 — APPROVED** [review]

File: src/components/mapChart/index.tsx (hunk 9 above)
Comment (VWSCoronaDashboard1) [reply src/components/mapChart/index.tsx:89]: This is what's in the official Highcharts typings ;)

Comment (VWSCoronaDashboard4) [reply src/components/mapChart/index.tsx:89]: if you'd supply the subset that could be returned by your function, it would be plenty? :P

**Review by VWSCoronaDashboard4 — APPROVED** [review]

File: src/components/mapChart/index.tsx (hunk 9 above)
Comment (VWSCoronaDashboard1) [reply src/components/mapChart/index.tsx:89]: fair enough, changed it

**Pull request merged**